server/src-lib/Hasura/GraphQL/Execute/LiveQuery/Poll.hs

```diff
@@ -64,7 +65,13 @@ data Subscriber
   , _sOnChangeCallback :: !OnChange
   }
 
-type OnChange = GQResponse -> IO ()
+-- live query onChange metadata, used for adding more extra analytics data
+data LiveQueryMetadata
+  = LiveQueryMetadata
+  { _lqmExecutionTime :: !(Maybe Double)
+  }
+
+type OnChange = GQResponse -> LiveQueryMetadata -> IO ()
```

Comment:
A `GQResponse` can be an error, and it seems unlikely that the `LiveQueryMetadata` information is very useful in that case. Might it make sense to instead use something like

```haskell
data LiveQueryResponse
  = LiveQueryResponse
  { _lqrPayload :: !BL.ByteString
  , _lqrExecutionTime :: !DiffTime
  }

type OnChange = GQResult LiveQueryResponse -> IO ()
```

instead?

Reply:
@lexi-lambda currently LiveQuery module are too coupling on `GQResponse` because it is content of `GraphqlResponse`. If we change the structure, it  will take many step too convert back and fore between `GQResult LiveQueryResponse` to `GQResponse`. Do you think we need to decouple live query `GraphqlResponse` too?

```haskell
data DataMsg
  = DataMsg
  { _dmId      :: !OperationId
  , _dmPayload :: !GraphqlResponse
  }

data GraphqlResponse
  = GRHasura !GQResponse
  | GRRemote !RemoteGqlResp
```

Reply:
I see—I misunderstood the way these types were being used. It seems like maybe `GQResponse` should just be updated to be `GQResult LiveQueryResponse`, then, since as far as I can tell, `GQResponse` is only used for websockets? But in general these types seem a little unnecessarily twisty to me; it might be better to do a little untangling to find a better representation.

I’m not immediately sure what that representation ought to be, but I think the end goal of avoiding including `LiveQueryMetadata` with errors seems reasonable. If you disagree, then by all means, I could be convinced to leave the type alone: I don’t feel super strongly about it. But if the current approach is just a side-effect of the types currently being structured the way they are rather than an intentional design decision, I think it would be nice to do something better.

Reply:
IMO, we can define a type class `EncoderJSON` to generalize `GQResult` and `encodeGQResp` function

```haskell
class EncoderJSON a where
  encJSON :: a -> EncJSON

-- encodeGQResp :: GQResponse -> EncJSON
-- change to
-- encodeGQResp :: EncoderJSON a => GQResult a -> EncJSON
```

We won't care about `GQSuccess` content anymore. Moreover, we can manage what fields are returned instead of encoding entire data